versioned_docs/version-5.x/faq/migrating-from-ink-4-to-5.md

```diff
@@ -0,0 +1,460 @@
+with a reserved/well known selector. This guarantees that there are no selector clashes,
+either by chance or malicious intent, and that the Proxy will only handle messages intended for it.
+
+If a contract uses a wildcard selector `#[ink(message, payable, selector = _)]` it _MAY_ define one
+other message. This message _MUST_ use the reserved selector `@`.
+This selector _MUST_ only be used in conjunction with a wildcard selector.
+
+```rust
+/// Handles any message with no matching selector in this proxy contract
+#[ink(message, selector = _)]
+pub fn fallback(&self) {
+    // forward call to the "logic" contract which actually executes the call
+}
+
+#[ink::scale_derive(Decode)]
+pub enum ProxyMessage {
+    UpgradeContract(Hash),
+}
+
+/// One other message allowed to handle messages.
+/// Fails to compile unless `@` is used as the selector.
+#[ink(message, selector = @)]
+pub fn handler(&self, msg: ProxyMessage) {
+  match msg {
+    ProxyMessage(hash) => { }
+  }
+}
+
+/// An additional message. Fails to compile when uncommented.
+// #[ink(message)]
+// pub fn additional_message(&self, msg: ProxyMessage) {
+//    match msg {
+//        ProxyMessage(hash) => ...
+//    }
+// }
+```
+
+### Events 2.0
+
+In prior ink! versions, events were defined inside the `#[ink::contract]` macro.
+With ink! 5.0 we decouple events from the `#[ink::contract]` macro,
+allowing events to be shared between contracts.
+We've updated [the Events documentation page](/5.x/basics/events) accordingly.
+
+The syntax of defining events within the main `#[ink::contract]` macro will continue to work,
+no code changes in existing contracts are required to update to the new syntax.
+
+:::caution
+The topic calculation changed in general, so also for events that are declared inside the
+`#[ink::contract]` macro!
+
+This is a breaking change for any client code which uses topics to filter events.
+
+Please see [#1827](https://github.com/paritytech/ink/pull/1827) for details.
+:::
+
+### No more unchecked arithmetic
+
+Unchecked arithmetic operations in a contract no longer work.
+Compiling a contract that contains those will fail gracefully.
+You need to handle overflows using e.g. saturating or wraparound math.
+
+This was done in [#1831](https://github.com/paritytech/ink/pull/1831).
+
+### Fail when decoding from storage and not all bytes consumed
+
+If a contract previously relied on successful decoding which does not consume all bytes,
+then recompiling with a version of ink! which includes this change will cause that contract
+to trap at runtime when attempting to decode.
+
+A simple example would be if a storage cell contains some bytes which were in the first place
+an encoded `u32`. If the contract attempts to decode those into a `u8`
+this would previously have succeeded, now the contract would trap.
+
+We introduced this change in [#1897](https://github.com/paritytech/ink/pull/1897).
+
+### [ink_e2e] Improved `call()` API
```

Comment:
Missing the notion of builder API in e2e: https://github.com/paritytech/ink/pull/1917